src/Domain/Item/ItemService.ts

```diff
@@ -49,8 +55,138 @@ export class ItemService implements ItemServiceInterface {
     }
   }
 
-  async saveItems(_itemHashes: string[], _userAgent: string, _retrievedItems: Item[]): Promise<SaveItemsResult> {
-    throw new Error('Method not implemented.')
+  async saveItems(dto: SaveItemsDTO): Promise<SaveItemsResult> {
+    const savedItems: Array<Item> = []
+    const conflicts: Array<ItemConflict> = []
+
+    const lastUpdatedTimestamp = this.timer.getTimestampInMicroseconds()
+
+    for (const itemHash of dto.itemHashes) {
+      const existingItem = await this.itemRepository.findByUuidAndUserUuid(itemHash.uuid, dto.userUuid)
+
+      if (!this.itemShouldBeSaved(itemHash, existingItem)) {
+        conflicts.push({
+          serverItem: existingItem,
+          type: 'sync_conflict'
+        })
+
+        continue
+      }
+
+      if(existingItem) {
+        const updatedItem = await this.updateExistingItem(existingItem, itemHash, dto.userAgent)
+        savedItems.push(updatedItem)
+      } else {
+        try {
+          const newItem = await this.saveNewItem(itemHash, dto.userAgent)
+          savedItems.push(newItem)
+        } catch (error) {
+          conflicts.push({
+            unsavedItem: itemHash,
+            type: 'uuid_conflict',
+          })
+
+          continue
+        }
+      }
+    }
+
+    const syncToken = this.calculateSyncToken(lastUpdatedTimestamp, savedItems)
+
+    return {
+      savedItems,
+      conflicts,
+      syncToken
+    }
+  }
+
+  async frontLoadKeysItemsToTop(userUuid: string, retrievedItems: Array<Item>): Promise<Array<Item>> {
+    const itemsKeys = await this.itemRepository.findAll({
+      userUuid,
+      contentType: Item.CONTENT_TYPE_ITEMS_KEYS,
+      sortBy: 'updated_at_timestamp',
+      sortOrder: 'DESC'
+    })
+
+    const retrievedItemsIds: Array<string> = retrievedItems.map((item: Item) => item.uuid)
+
+    itemsKeys.forEach((itemKey: Item) => {
+      if(retrievedItemsIds.indexOf(itemKey.uuid) === -1) {
+        retrievedItems.unshift(itemKey)
+      }
+    })
+
+    return retrievedItems
+  }
+
+  private calculateSyncToken(lastUpdatedTimestamp: number, savedItems: Array<Item>): string {
+    if (savedItems.length) {
+      const sortedItems = savedItems.sort((itemA: Item, itemB: Item) => itemA.updatedAt > itemB.updatedAt ? 1 : -1)
+      lastUpdatedTimestamp = sortedItems[sortedItems.length - 1].updatedAt
+    }
+
+    const lastUpdatedTimestampWithMicrosecondPreventingSyncDoubles = lastUpdatedTimestamp + 1
+
+    return Buffer.from(
+      `${this.SYNC_TOKEN_VERSION}:${lastUpdatedTimestampWithMicrosecondPreventingSyncDoubles / 1000000}`,
```

Comment:
Do you think it would be safer/more readable to extract `1000000` into a const called `NumMicrosecondsInSeconds`? Also to use underscores in large numbers to improve readability: `1_000_000`. I keep having to count the zeros to make sure :)

Reply:
Sounds good. Will fix 👍